app/src/main/java/com/kpiroom/bubble/ui/login/LoginLogic.kt

```diff
@@ -1,22 +1,138 @@
 package com.kpiroom.bubble.ui.login
 
+import android.util.Log
 import androidx.lifecycle.MutableLiveData
+import com.kpiroom.bubble.source.Source
+import com.kpiroom.bubble.source.api.impl.firebase.FirebaseAuthUtil
+import com.kpiroom.bubble.source.api.impl.firebase.FirebaseExceptionHelper
 import com.kpiroom.bubble.ui.core.CoreLogic
+import com.kpiroom.bubble.util.async.AsyncBag
+import com.kpiroom.bubble.util.async.AsyncProcessor
+import com.kpiroom.bubble.util.constant.str
+import com.kpiroom.bubble.util.databinding.ProgressState
+import com.kpiroom.bubble.util.databinding.ProgressState.*
+import com.kpiroom.bubble.util.databinding.ProgressStateContainer
+import com.kpiroom.bubble.R
 
-class LoginLogic: CoreLogic() {
+class LoginLogic : CoreLogic() {
 
-    val login = MutableLiveData<String>()
+    companion object {
+        const val TAG = "LoginLogic"
+    }
+
+    val email = MutableLiveData<String>()
     val password = MutableLiveData<String>()
+    val confirmPassword = MutableLiveData<String>()
 
-    private fun signUpOrSignIn() {
-        //TODO("Implement db request for signin or signup")
-        //login.value
-        //password.value
+    val authButtonText = MutableLiveData<String>().setDefault(str(R.string.login_sign_in))
+    val changeAuthButtonText = MutableLiveData<String>().setDefault(str(R.string.login_sign_up))
+    val isNewAccount = MutableLiveData<Boolean>()
+
+    val delayedAction = MutableLiveData<DelayedAction>()
+    val authProgressState = MutableLiveData<ProgressStateContainer>()
+
+    val api = Source.api
+
+    fun onAuthClicked() {
+        clickThrottler.next {
+            delayedAction.value = DelayedAction({
+                signUpOrSignIn()
+            }, 300)
+        }
     }
 
-    fun onGetStarted() {
+    fun onForgotPassword() {
         clickThrottler.next {
-            signUpOrSignIn()
+            val email = email.value
+
+            if (email.isNullOrEmpty()) {
+                authProgressState.setValue(ALERT, str(R.string.message_forgot_password_no_email))
+                return@next
+            }
+
+
+            AsyncProcessor {
+                api.sendPasswordResetEmail(email)
+                authProgressState.postValue(
+                    ALERT,
+                    str(R.string.message_forgot_password_instructions_sent)
+                )
+            }.handleError {
+                authProgressState.postValue(ALERT, "${str(R.string.message_forgot_password_error)}: $it")
+            }.run(bag)
+        }
+    }
+
+    fun toggleNewAccount() {
+        swapValues(authButtonText, changeAuthButtonText)
+
+        val newValue = isNewAccount.value != true
+        if (!newValue) {
+            delayedAction.value = DelayedAction({ confirmPassword.value = "" }, 300)
         }
+        isNewAccount.value = newValue
     }
+
+    private fun signUpOrSignIn() {
+
+        authProgressState.setValue(LOADING)
```

Comment:
rename 'authProgressState' to just 'progress' and add extension like
```
fun MutableLiveData<ProgressStateContainer>.load() {
setValue(LOADING)
}
```
And use progress.load()